File: problems/0042-trapping-rain-water/typescript/solution.ts
// Trapping Rain Water — https://leetcode.com/problems/trapping-rain-water/
export function trap(height: number[]): number {
  if (height.length === 0) return 0;
  let left = 0;
  let right = height.length - 1;
  let leftMax = height[left];
  let rightMax = height[right];
  let water = 0;
  while (left < right) {
    if (leftMax < rightMax) {
      left++;
      leftMax = Math.max(leftMax, height[left]);
      water += leftMax - height[left];
    } else {
      right--;
      rightMax = Math.max(rightMax, height[right]);
      water += rightMax - height[right];
    }
  }
  return water;
}

File: problems/0042-trapping-rain-water/go/solution.go
package solution

// Trapping Rain Water — https://leetcode.com/problems/trapping-rain-water/
func trap(height []int) int {
	if len(height) == 0 {
		return 0
	}
	left, right := 0, len(height)-1
	leftMax, rightMax := height[left], height[right]
	water := 0
	for left < right {
		if leftMax < rightMax {
			left++
			if height[left] > leftMax {
				leftMax = height[left]
			}
			water += leftMax - height[left]
		} else {
			right--
			if height[right] > rightMax {
				rightMax = height[right]
			}
			water += rightMax - height[right]
		}
	}
	return water
}
